CI: Adding script to validate consistent and correct capitalization among headings in documentation (#26941)

tonywu1999 opened the pull request:
Closes #26941

This script should be running correctly to validate capitalization for various rst files.  Many keywords will need to be added into the script as more and more files are tested on this script. Currently testing on the following rst files through code_checks.sh:

doc/source/development/contributing.rst
doc/source/index.rst doc/source/ecosystem.rst

contributing.rst and ecosystem.rst should produce an exit code of 1, indicating there are incorrect capitalization in their headings.

Commit:
changed name

Commit:
adding sphinx extension

Commit:
Starting builder

Commit:
experimenting with builder

Commit:
before running build

Commit:
126 warnings?

Commit:
contributors

Commit:
experimenting

Commit:
update

Commit:
parser created

Commit:
italics working

Commit:
found a way to collect all heading strings from doctree

Commit:
testing script

Commit:
modified validation script

Commit:
command line arguments possible for validation script

Commit:
validation script needs better commenting

Commit:
added line number to validation script

Commit:
edited code_checks.sh

Commit:
argument parser correctly implemented

Commit:
Added comments

Commit:
Validate consistency of title capitalization in documentation script added (#26941)

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

Commit:
Adding script to validate consistency of title capitalization (#26941)

Commit:
Adding validate_rst_title_capitalization.py (#26941)

pep8speaks commented:
Hello @tonywu1999! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`scripts/validate_rst_title_capitalization.py`](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py):

> [Line 14:89](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L14): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 88 characters)
> [Line 36:1](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L36): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 55:24](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L55): [E222](https://duckduckgo.com/?q=pep8%20E222) multiple spaces after operator
> [Line 55:45](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L55): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 61:33](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L61): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 62:33](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L62): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 66:33](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L66): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 67:33](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L67): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 75:34](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L75): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 97:1](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L97): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 107:49](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L107): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 107:89](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L107): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 88 characters)
> [Line 140:1](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L140): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 152:21](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L152): [E711](https://duckduckgo.com/?q=pep8%20E711) comparison to None should be 'if cond is None:'
> [Line 157:1](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L157): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 172:9](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L172): [E123](https://duckduckgo.com/?q=pep8%20E123) closing bracket does not match indentation of opening bracket's line
> [Line 177:89](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L177): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 88 characters)
> [Line 184:1](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L184): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 184:89](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L184): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 88 characters)
> [Line 195:89](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L195): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 88 characters)
> [Line 206:30](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L206): [W605](https://duckduckgo.com/?q=pep8%20W605) invalid escape sequence '\ '
> [Line 228:89](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L228): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 88 characters)
> [Line 242:89](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L242): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 88 characters)
> [Line 261:13](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L261): [E125](https://duckduckgo.com/?q=pep8%20E125) continuation line with same indent as next logical line
> [Line 261:13](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L261): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 266:89](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L266): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 88 characters)
> [Line 275:89](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L275): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 88 characters)
> [Line 283:1](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L283): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 301:89](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L301): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (90 > 88 characters)
> [Line 304:89](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L304): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 88 characters)
> [Line 322:89](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L322): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 88 characters)
> [Line 339:1](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L339): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 360:89](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L360): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 88 characters)
> [Line 367:49](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L367): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 367:51](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L367): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 367:89](https://github.com/pandas-dev/pandas/blob/635163d783e7c2200389697f967e01172ce4dda8/scripts/validate_rst_title_capitalization.py#L367): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 88 characters)

WillAyd commented:
This is a lot of code; is there no third party package to do something like this?

Commit:
Testing validate_rst_capitalization.py script (#26941)

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

tonywu1999 commented:
@WillAyd 

I've looked around when trying to solve this problem and I don't think there's a third party package that does everything that we need.  Ultimately, the more complicated parts of the script involve the combination of finding headers in an RST file & determining if a string follows our capitalization convention while ensuring keywords like pandas are properly capitalized too.

I will keep looking for third party packages online to hopefully simplify the lines of this script.  In the meantime, my goal right now is to check if the script is working as expected.

Commit:
Edited validate script (#26941)

datapythonista commented on ci/code_checks.sh:
```diff
@@ -318,6 +318,11 @@ if [[ -z "$CHECK" || "$CHECK" == "docstrings" ]]; then
     $BASE_DIR/scripts/validate_docstrings.py --format=actions --errors=GL03,GL04,GL05,GL06,GL07,GL09,GL10,SS04,SS05,PR03,PR04,PR05,PR10,EX04,RT01,RT04,RT05,SA02,SA03,SA05
     RET=$(($RET + $?)) ; echo $MSG "DONE"
 
+    ### VALIDATE TITLE CAPITALIZATION CONVENTION ###
```
I think this comment is redundant with the message in next line, you can probably get rid of it

datapythonista commented on ci/code_checks.sh:
```diff
@@ -340,5 +345,4 @@ if [[ -z "$CHECK" || "$CHECK" == "typing" ]]; then
     RET=$(($RET + $?)) ; echo $MSG "DONE"
 fi
 
-
```
Try to avoid unrelated changes, even if they're small

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,394 @@
+    'pandas', 'Python', 'IPython', 'PyTables', 'Excel', 'JSON',
+    'HTML', 'SAS', 'SQL', 'BigQuery', 'STATA', 'Interval', 'PEP8',
+    'Period', 'Series', 'Index', 'DataFrame', 'C', 'Git', 'GitHub', 'NumPy',
+    'Apache', 'Arrow', 'Parquet', 'Triage', 'MultiIndex', 'NumFOCUS', 'sklearn-pandas'
```
I don't think triage should be capitalized, I can be wrong

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,394 @@
+
+# Dictionary of bad titles that will be printed later along with line numbers
+# Key: Document Directory, Value: Pair(Bad Title, Line Number)
+badTitleDictionary = {}
```
In Python the convention is to use `bad_title_dictionary` instead of camel case. This applies to all variables, and function names in your code. Have a look at the PEP-8 style guide.

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,394 @@
+
+    '''
+    def __init__(self):
+        # Open a pair of null files
```
Better avoid the comments where you explain the code, python code is very readable. Not the init, but for the other functions of you add a comment explaining what's their goal, use a docstring instead of a comment with a hash

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,394 @@
+                        f.append(os.path.join(dirpath, file))
+
+    # Return the filled up list of .rst file paths
+    return f
```
I think it's probably better to use a generator for this. Also, I'd avoid making the function "create", I think it's misleading (like creating a directory in disk...)

datapythonista reviewed (commented):
Nice work. You probably want to check the other scripts, you'd see were very consistent with styles, and you'll need to adapt. But looks good, the context manager to supress the output is probably the best option.

Thanks!

Commit:
Added parameter and return value information in docstrings

Commit:
Edited validate_rst_title_capitalization.py for review (#29641)

Commit:
Edited validate_rst_title_capitalization.py for review (#26941)

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

tonywu1999 commented:
I'm seeing some problems in the CI/Checks that are occurring outside of the errors I purposely created myself.  I'm not exactly sure if my script is causing these errors or if these are errors independent of my script.

1. Linting
Check import format using isort
##[error]Skipped 2 files
Check import format using isort DONE
##[error]Process completed with exit code 1

2. Publish benchmarks artifacts
##[error]Path does not exist /home/runner/work/pandas/pandas/asv_bench/benchmarks.log
##[error]Exit code 1 returned from process: file name '/home/runner/runners/2.163.1/bin/Runner.PluginHost', arguments 'action "GitHub.Runner.Plugins.Artifact.PublishArtifact, Runner.Plugins"'.

Furthermore, I'm having trouble suppressing stderr output that is produced from an imported helper method.  When I attempt to suppress the stderr output on my local machine using a context manager (_class suppress_stdout_stderr_ on line 35), the suppressing works.  However, only when I push my code onto GitHub does the suppressing fail to work on CI/Checks.  This causes more clutter than I would want to appear on CI/Checks.

datapythonista commented:
I'm in my phone and can't access github actions, but I think both errors are false positives that fail in all PRs, but don't make the CI fail. Is it possible that your script is not exiting with a zero code, or that another error is happening?

tonywu1999 commented:
Ok, so to reiterate what you said, I do not need to worry about the 'linting' and 'publish benchmark artifacts' checks.  

Also, I set an exit code of 1 in validate_rst_title_capitalization.py if there are titles/headings that have improper capitalization.  Currently, the script is checking to validate a file that has headings with improper capitalization, so it is returning an exit code of 1 at the moment.  Is this the reason why the context manager is able to suppress stderr output on my local machine but not on GitHub?  I'm very confused on what is going on with regards to suppressing output.

datapythonista commented:
Just ignore those errors then.

I don't think what I said is related to the output not being suppressed in github, not sure about that.

MomIsBestFriend commented:
@tonywu1999 thank you for the code.

You'll need to run "black" on the new created file.

```bash
$ black  path/to/your/file
```

(Don't panic, but it's going to change a lot of lines).

MomIsBestFriend commented:
> I'm in my phone and can't access github actions, but I think both errors are false positives that fail in all PRs, but don't make the CI fail. Is it possible that your script is not exiting with a zero code, or that another error is happening?

```Options -> Request Desktop site``` :)

cc @datapythonista

Commit:
Checking if stderr output will be suppressed (#26941)

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

Commit:
Simplified validate_rst_title_capitalization.py to print correctly (#26941)

tonywu1999 commented:
@datapythonista 

Because the suppressing wasn't working, I changed my code so that rather than using a third party module to find headings in an RST file, I replaced the module with my own code from scratch.  Now, I'm fairly confident the script is outputting on CI/Checks exactly what I want it to output.  

Where should I go from here?  Also if a file contains headings that are not properly capitalized, should I return an system exit code other than zero?

I also noticed a new error in 'continuous-integration/travis-ci/pr', and I'm not sure if that is related to anything in my script.

Commit:
Testing script on doc/source/development/contributing.rst (#26941)

MomIsBestFriend commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,306 @@
+err_msg = "Heading capitalization formatted incorrectly. Please correctly capitalize"
+
+
+def follow_capitalization_convention(title: str) -> bool:
```
Since this is returning a bool, I think that ```is_following_capitalization_convention``` is more appropriate function name.

MomIsBestFriend commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,306 @@
+    # Open and read the .rst file and store the string of data into input
+    f = open(rst_file, "r")
+    input = f.read().split("\n")
+    f.close()
```
Can you use a contex manager here?

I believe that

```python
with open(rst_file, "r") as file_obj:
    lines = file_obj.read().split("\n")
```

Should do the trick here.

side note, ```input``` is a built-in python function, please try to avoid using it as a variable name

MomIsBestFriend commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,306 @@
+                bad_title_dict[rst_file].append((title_list[i], line_number_list[i]))
+
+
+def find_rst_files(source_paths: List[str]) -> List[str]:
```
I think that the return type here is:

```python
Generator[str, None, None]
```

tonywu1999 replied:
@MomIsBestFriend Thank you for your input!  It's really helping me a lot.

MomIsBestFriend replied:
No problems, and if I gotta be honest I learnt a lot from @datapythonista, I would suggest you to read his code suggestions he gave me in https://github.com/pandas-dev/pandas/pull/30467 and in https://github.com/pandas-dev/pandas/pull/30755

I think that I stepped in pretty much every possible pitfall there is.

Commit:
validate_rst_title_capitalization.py MomIsBestFriend edits (#26941)

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,307 @@
+# Key: Document Directory, Value: Pair(Bad Title, Line Number)
+bad_title_dict = {}
+
+# Error Message:
```
Please delete

tonywu1999 replied:
Could you clarify what I should be deleting here?

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,307 @@
+    word_list = re.split(r"[;,-/():\s]\s*", title)
+
+    # Edge Case: First word is an empty string
+    if len(word_list[0]) == 0:
```
`if not word_list[0]:`

Do you know why is this happening?

tonywu1999 replied:
I came across situations where the text in a heading was like the following:

'  heading'

where there was some number of spaces in the beginning of the heading.  In these situations, the split function would cause the first word in the word_list array to be an empty string.

tonywu1999 replied:
In this situation, should the heading be considered incorrect?

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,307 @@
+        return False
+
+    # Dealing with the first word of the title
+    if word_list[0] not in CAPITALIZATION_EXCEPTIONS:
```
I think this should be part of the next loop, inside the loop check first if the word is a proper name, which is common for all words, and then have an if to see if it's the first word or not (you can use `enumerate` in the loop for it)

datapythonista replied:
Or better implementation could be to have a function `correct_title_capitlization(title), that simply calls the string function to make the first letter upper case and ask the rest lower case, and then replace proper words of any from the list.

This way you can add in the error message what is the correct capitalization.

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,307 @@
+
+    """
+
+    is_failed: bool = False
```
I would count how many errors are there and return the counter as the exit code

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,307 @@
+    return True
+
+
+def findTitles(rst_file: str) -> Generator[List[str], List[int], None]:
```
Avoid camel case

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,307 @@
+    title_list : List[str]
+        A list of heading strings found in the document tree
+
+    line_number_list : List[int]
```
Return tuples of title de longe number instead of two lists

datapythonista reviewed (commented):
Nice work, added some more comments X but this is looking much better. Please avoid comments explaining the code, unless it's not possible that the code is clear enough (which happens very rarely in Python).

But this starts to look much better.

The exit code, as I said in a comment, my preference is returning the number of errors found.

Thanks a lot for working on this!

MomIsBestFriend commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,307 @@
+    is_failed = True
+    for key in bad_title_dict:
+        for line in bad_title_dict[key]:
+            print(key + ":" + str(line[1]) + ": " + err_msg + ' "' + line[0] + '"')
```
```python
print(f'{key}:{line[1]}:{err_msg} "{line[0]"')
```

MomIsBestFriend commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,307 @@
+        elif directory_address.endswith(".rst"):
+            yield directory_address
+        else:
+            for (dirpath, dirnames, filenames) in walk(directory_address):
```
```dirnames``` never gets used, can you replace it with ```_```?

should like something like this:

```python
for dirpath, _, filenames in os.walk(directory_address):
```

Commit:
Created method to correct title capitalization (#26941)

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

Commit:
Ran black on validate_rst_title_capitalization (#26941)

Commit:
Edit: titles with non-word character as first character are not valid

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

tonywu1999 commented:
I edited the script so that now there is a function (correct_title_capitalization) that let's the user know what the corrected capitalization title should be.

Example output:
ci/../doc/source/development/contributing.rst:824:Heading capitalization formatted incorrectly. Please correctly capitalize "Type Hints" to "Type hints"

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,272 @@
+    directory_list = find_rst_files(source_paths)
+
+    for filename in directory_list:
+        fill_bad_title_dict(filename)
```
This F feels overcomplicated. Just loop over `find_titles` here, and if `title == correct_title_capitalization(title)` yield a tuple with the file name, and the correct and incorrect titles.

I think you can get rid a lot of the code by doing that, and things will be much clearer.

MomIsBestFriend commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,272 @@
+    for word in word_list:
+        if word.lower() in CAP_EXCEPTIONS_DICT:
+            correct_title = re.sub(
+                r"\b" + word + r"\b", CAP_EXCEPTIONS_DICT[word.lower()], correct_title
```
Can you please use f-strings here?

```python
correct_title = re.sub(rf"\b{word}\b", CAP_EXCEPTIONS_DICT[word.lower()], correct_title)
```

Commit:
Simplified validate_rst_title_capitalization main method (#26941)

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

Commit:
Edited parameter and return value description of main function

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

tonywu1999 commented:
I edited the script so the main function has been simplified.  I got rid of the bad_title_dictionary as it is not needed anymore.

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

tonywu1999 commented:
@datapythonista 

Are there any additional steps I should take to get the pull request approved?  I'm not sure where to go from here.

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,217 @@
+
+    table = str.maketrans("", "", "*`_")
+
+    for line_no in range(1, len(lines)):
```
I think it's more readable to use `for i, line in enumerate(lines):` and then access the previous line with `lines[i - 1]`.

Using regex also is probably overcomplicating things. You can do something like:
```python
line_chars = set(line)
if len(line_chars) == 1 and line_chats.pop() in ('=', '*',...) and len(line) == len(line[i - 1]):
  ... 
```

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,217 @@
+        elif directory_address.endswith(".rst"):
+            yield directory_address
+        else:
+            for (dirpath, _, filenames) in os.walk(directory_address):
```
Can you check if the `glob` module can make this easier?

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,217 @@
+
+    """
+
+    correct_title: str = re.sub(r"^\W*", "", title).capitalize()
```
Can you add a comment here, and may be the next two lines, to explain what's going on? It's not immediate to understand what those lines are doing.

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,217 @@
+    Parameters
+    ----------
+    title : str
+        Heading string to correct
```
Can you finish this sentence and the description of returns with a period please.

datapythonista reviewed (commented):
Thanks, looks great. Just one section feels a bit overcomplicated, but other than that looks really great.

Commit:
Added glob module to script 01-27-2020

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

tonywu1999 commented:
@datapythonista

Thank you for your input!  I made the changes that you requested 3 days ago.

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,214 @@
+    table = str.maketrans("", "", "*`_")
+
+    for i, line in enumerate(lines):
+        if len(line) != 0 and len(lines[i - 1]) != 0:
```
`if len(line) != 0` should be simply `if line` (both cases)

tonywu1999 replied:
@datapythonista made the correction that you requested.  Thank you for your help in reviewing my pull request

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,214 @@
+    for i, line in enumerate(lines):
+        if len(line) != 0 and len(lines[i - 1]) != 0:
+            line_chars = set(line)
+            if len(line_chars) == 1 and line_chars.pop() in symbols:
```
Why is not as simple as checking that besides this condition the line and the previous have the same length?

I guess I'm missing something, but can't see why all next lines are required, sorry.

tonywu1999 replied:
I read the RST documentation guide on what is considered a heading, and there are two situations. 
Case 1:
The heading is between two lines like so:
*******
heading
*******
In this situation, the heading has the be the same length as the length of the line below it.

Case 2:
Heading
***************
In this case, there are no stars above the heading.  In this situation, the heading length only needs to be less than or equal to the length of the line below it to be considered a heading in rST

datapythonista replied:
I think we always use case 2, with the same length in both lines. I'd personally ignore all options of rst and make the script simpler. What do you think?

tonywu1999 replied:
Ok that sounds good to me.  I'll make the changes soon

tonywu1999 replied:
@datapythonista made the appropriate changes.

datapythonista reviewed (approved):
A question and a small style comment, but lgtm, thanks

Commit:
Edited len(line) != 0 correction

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

Commit:
Edited find_titles method

tonywu1999 commented:
Thank you for your guidance @datapythonista

datapythonista commented:
@pandas-dev/pandas-core CI is red to show how the errors look like, if someone can review we can fix the errors and merge.

WillAyd commented:
I'm +/- 0 on this. I don't know that the errors highlighted really are all errors, nor that its worth taking this on in pandas. 

But don't oppose strongly if others feel this is moving in the right direction

datapythonista commented:
@tonywu1999 can you add to the description the issue you're addressing here please? @WillAyd we've got some discussion about this, and one of the main issues is that even in the navigation we have inconsistencies, like `pandas code style` and `Pandas Maintenance` next to each other.

The agreement was to not capitalize more than the first letter, and fix al titles.

tonywu1999 commented:
@datapythonista when you say description, do you mean in the first comment block of my code?  Also do you want me to simply add "GH 26941" to the description?

datapythonista commented:
I mean in this github issue, not in the code, sorry

tonywu1999 commented:
@datapythonista I changed the description of this issue.  Let me know if you need anything else!

datapythonista commented:
That is the title. ;)

I updated the description myself. This gives context to reviewers who otherwise think there has been no previous discussion. And also, this will make github close the issue once this is merged.

tonywu1999 commented:
@datapythonista Haha my bad.  Thank you for your help.

jbrockmendel commented:
@tonywu1999 looks like some (maybe new?) docstrings dont pass this check.  can you update

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

datapythonista commented:
@tonywu1999 can you fix the titles you left to see how the errors look like? I think we can get this merged once it's green.

datapythonista commented:
Also, please merge master into your branch, so we make sure we're validating the latest version of the docs. Thanks!

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

Commit:
edited contributing.rst to have no errors

tonywu1999 commented:
Hi @datapythonista 

I fixed the titles just now and committed the changes.  There should be title changes in doc/source/development/contributing.rst

datapythonista reviewed (approved):
Thanks, can you have a look at the CI? I guess it's unrelated, but if you can merge master and make sure it's green, well be able to merge this.

tonywu1999 commented:
@datapythonista 
I'm not sure what you meant by that.  I merged with master and CI/Checks, which my code affects, is running fine.  Are you referring to the two X's for pandas-dev.pandas and pandas-dev.pandas(Linux py36_locale) ?

datapythonista commented:
Yes, something failed in that build. Probably something unrelated to this pr, but we need to have it green to merge this. I'm in my phone and can't look at the logs right now, but if you can have a look and see what's the problem, that would be great.

tonywu1999 commented:
@datapythonista 

I looked at the error logs. However, I'm not really sure what's causing the errors.

datapythonista commented:
> I looked at the error logs. However, I'm not really sure what's causing the errors.

If you navigate all the way to the azure pipeline logs, and look for errors, you'll find this line:
```
AssertionError: Caused unexpected warning(s): [('ResourceWarning', ResourceWarning('unclosed <ssl.SSLSocket [closed] fd=16, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM,
```

Which looks like a systems error, that (hopefully) shouldn't repeat if the job is executed again. If you merge master into your branch, and push, the CI will run again, and hopefully everything will be green this time. And we'll be able to merge this with confidence.

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

tonywu1999 commented:
@datapythonista 

I think all checks have passed now.

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,211 @@
+
+    """
+
+    # Function to strip all non-word characters from the beginning of the title to the
```
Do you mind removing the "Function to" part of all those? I find it a bit confusing, since it's not a function.

Just the "Strip all non-word..." should be enough.

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,211 @@
+    with open(rst_file, "r") as file_obj:
+        lines = file_obj.read().split("\n")
+
+    symbols = ("*", "=", "-", "^", "~", "#", '"')
```
Probably worth making this a constant and moving it to the top of the file with the list of words with fiixed capitalization.

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,211 @@
+            and line_chars.pop() in symbols
+            and len(line) == len(lines[i - 1])
+        ):
+            yield lines[i - 1].translate(table), i
```
I'd personally prefer to use:
```python
yield re.sub('[`\*_]', '', lines[i - 1]), i
```
Should be the same, but I think `str.translate` is rarely used and not many people knows about it, and using the regex equivalent should be easier to read for most people.

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,211 @@
+    """
+
+    with open(rst_file, "r") as file_obj:
+        lines = file_obj.read().split("\n")
```
I think the standard way to do this would be:
```python
with open(rst_file, "r") as fd:
    for i, line in enumerate(fd):
     ....
```

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,211 @@
+
+    Yields
+    -------
+    directory_address : str
```
```suggestion
    str
```

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,211 @@
+    -------
+    directory_address : str
+        Directory address of a .rst files found in command line argument directories.
+
```
Better remove this blank line (we're not validating this docstring, but this blank line would make it fail if we do.

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,211 @@
+
+    Returns
+    -------
+    correct_title : str
```
```suggestion
    str
```

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,211 @@
+
+CAP_EXCEPTIONS_DICT = {word.lower(): word for word in CAPITALIZATION_EXCEPTIONS}
+
+err_msg = "Heading capitalization formatted incorrectly. Please correctly capitalize"
```
Since we're using this just once, I'd move it to the function where it's being used.

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,211 @@
+    -------
+    title : str
+        A heading found in the rst file.
+
```
No blank line here

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,211 @@
+
+    line_number : int
+        The corresponding line number of the heading.
+
```
No blank line here either.

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,211 @@
+        Directory address of a .rst files found in command line argument directories.
+
+    """
+
```
No blank line.

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,211 @@
+        Number of incorrect headings found overall.
+
+    """
+
```
No blank line

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,211 @@
+
+    Returns
+    -------
+    number_of_errors : int
```
```suggestion
    int
```

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,211 @@
+    -------
+    number_of_errors : int
+        Number of incorrect headings found overall.
+
```
No blank line

datapythonista reviewed (commented):
Thanks @tonywu1999, and sorry for the delay.

I added few minor final comments more if you don't mind, I think they'll improve readability and future maintenance.

Commit:
modified validation script

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

Commit:
fix linting errors

Commit:
black pandas-dev change

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,206 @@
+    # first word character.
+    correct_title: str = re.sub(r"^\W*", "", title).capitalize()
+
+    # Function to remove a URL from the title. We do this because words in a URL must
```
Can you also remove the "Function to" from here.

datapythonista commented on scripts/validate_rst_title_capitalization.py:
```diff
@@ -0,0 +1,206 @@
+    # stay lowercase, even if they are a capitalization exception.
+    removed_https_title = re.sub(r"<https?:\/\/.*[\r\n]*>", "", correct_title)
+
+    # Function to split a title into a list using non-word character delimiters.
```
And from here.

Sorry, I thought I added it to the comment in the first occurrence.

datapythonista reviewed (commented):
I see there are still some pending comments from the previous review, let me know when this is ready to be reviewed again.

Commit:
modified changes to comments

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

Commit:
Merge remote-tracking branch 'upstream/master' into new-feature

datapythonista commented:
Thanks for this @tonywu1999, good improvement.

Can you please open an issue to fix the titles in the rest of the files?

tonywu1999 commented:
@datapythonista 

Thanks for your guidance! I opened a new issue (#32550) to address fixing the titles in the rest of the files
